File: scripts/command_benchmark.py
import socket
import time

server_ip = '127.0.0.1'
server_port = 8743
num_requests = 100
command = ''

latencies = []

def benchmark():
    client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    client_socket.connect((server_ip, server_port))
    sock_file = client_socket.makefile('r')

    try:
        for i in range(num_requests):
            start_time = time.time()

            message = command + "\n"
            client_socket.sendall(message.encode())

            response = sock_file.readline().strip()

            end_time = time.time()
            latency = end_time - start_time
            latencies.append(latency)

            print(f"[{i+1}] Response: {response}, Latency: {latency:.6f} sec")

    finally:
        client_socket.close()

    if latencies:
        print("\n=== Benchmark Results ===")
        print(f"Total requests: {num_requests}")
        print(f"Average latency: {sum(latencies)/len(latencies):.6f} sec")
        print(f"Max latency: {max(latencies):.6f} sec")
        print(f"Min latency: {min(latencies):.6f} sec")
        print(f"Throughput: {num_requests / sum(latencies):.2f} requests/sec")

if __name__ == "__main__":
    benchmark()

File: scripts/command_benchmark.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"sync"
	"time"
)

const (
	serverIP    = "127.0.0.1"
	serverPort  = "9000"
	numRequests = 100000
	numWorkers  = 100
	command     = "instruction=get|store=user_data|key=5"
)

func main() {
	results := make(chan float64, numRequests)
	startBenchmark := time.Now()

	var wg sync.WaitGroup
	for w := 0; w < numWorkers; w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()

			conn, err := net.Dial("tcp", serverIP+":"+serverPort)
			if err != nil {
				fmt.Println("Connection error:", err)
				return
			}
			defer conn.Close()

			reader := bufio.NewReader(conn)

			for i := 0; i < numRequests/numWorkers; i++ {
				start := time.Now()

				_, err := fmt.Fprintf(conn, "%s\n", command)
				if err != nil {
					fmt.Println("Send error:", err)
					return
				}

				_, err = reader.ReadString('\n')
				if err != nil {
					fmt.Println("Read error:", err)
					return
				}

				results <- time.Since(start).Seconds()
			}
		}()
	}

	wg.Wait()
	close(results)

	var latencies []float64
	for l := range results {
		latencies = append(latencies, l)
	}

	if len(latencies) == 0 {
		fmt.Println("No successful requests, exiting benchmark.")
		return
	}

	total := 0.0
	maxLatency := 0.0
	minLatency := 1e9
	for _, l := range latencies {
		total += l
		if l > maxLatency {
			maxLatency = l
		}
		if l < minLatency {
			minLatency = l
		}
	}

	elapsed := time.Since(startBenchmark).Seconds()

	fmt.Println("\n=== Benchmark Results ===")
	fmt.Printf("Total requests:   %d\n", len(latencies))
	fmt.Printf("Workers:          %d\n", numWorkers)
	fmt.Printf("Average latency:  %.6f sec\n", total/float64(len(latencies)))
	fmt.Printf("Min latency:      %.6f sec\n", minLatency)
	fmt.Printf("Max latency:      %.6f sec\n", maxLatency)
	fmt.Printf("Throughput:       %.2f req/sec\n", float64(len(latencies))/elapsed)
	fmt.Printf("Total time:       %.6f sec\n", elapsed)
}
